DL3062: relax DL3062, fix readme

m-ildefons opened the pull request:
DL3062 is very eager and sometimes triggers in cases where it doesn't make a whole lot of sense.
To fix this, it now doesn't trigger anymore when installing go packages from local filesystem paths. It also doesn't trigger anymore on command line arguments to go programs that are executed with `go run package@v1.2.3 arg1 arg2`

related-to: hadolint/hadolint#1130

m-ildefons committed:
DL3062: relax DL3062, fix readme

DL3062 is very eager and sometimes triggers in cases where it doesn't
make a whole lot of sense.
To fix this, it now doesn't trigger anymore when installing go packages
from local filesystem paths. It also doesn't trigger anymore on command
line arguments to go programs that are executed with `go run
package@v1.2.3 arg1 arg2`

related-to: hadolint/hadolint#1130

Signed-off-by: Moritz Röhrich <moritz@ildefons.de>

larry-c-ef reviewed (commented):
Thanks for tagging me. I don't speak enough Haskell to approve or disapprove, but I commented on the part I know about.

m-ildefons commented on README.md:
```diff
@@ -532,7 +532,7 @@ Please [create an issue][] if you have an idea for a good rule.
 | [DL3060](https://github.com/hadolint/hadolint/wiki/DL3060)   | Info             | `yarn cache clean` missing after `yarn install` was run.                                                                                            |
 | [DL3061](https://github.com/hadolint/hadolint/wiki/DL3061)   | Error            | Invalid instruction order. Dockerfile must begin with `FROM`, `ARG` or comment.                                                                     |
-| [DL3062](https://github.com/hadolint/hadolint/wiki/DL3061)   | Warning          | Pin versions in go install. Instead of `go install <package>` use `go install <package>@<version>`                                                  |
+| [DL3062](https://github.com/hadolint/hadolint/wiki/DL3062)   | Warning          | Pin versions in go install. Instead of `go install <package>` use `go install <package>@<version>`                                                  |
```
Thanks for reminding me of that wiki page. I've added it with a variety of example cases, which should make it clear that `go install`, `go get` and `go run` all fall under this rule.

m-ildefons committed:
dl3062: deal with tool commands

DL3062: Deal with `go ... tool` commands. Since Golang >= 1.24 the
package tooling for Golang can also manage tools as dependencies.
This introduces variations for the `go install` and `go get` commands.
The additional `tool` and `-tool` arguments respectively have to be
accepted by DL3062 and shall not lead to false positives.

related-to: hadolint/hadolint#1132

Signed-off-by: Moritz Röhrich <moritz@ildefons.de>

medasx commented on test/Hadolint/Rule/DL3062Spec.hs:
```diff
@@ -19,6 +19,28 @@ spec = do
+
+      describe "go install tool" $ do
+        it "version not pinned" $ do
+          ruleCatches "DL3062" "RUN go install tool foobar"
```
**issue**: go install treats `tool` argument as package. Only `RUN go install tool` needs to have special handling, as it will install all tools already present in go.mod. 
```
❯ go install tool golang.org/x/tools/cmd/stringer
go: warning: "tool" matched no packages
no required module provides package golang.org/x/tools/cmd/stringer; to add it:
	go get golang.org/x/tools/cmd/stringer
```
Note that tools are installed with `go get -tool [package]` which is already covered by the tests below.

m-ildefons replied:
Thanks for highlighting this. I've removed those test cases.
